### ISSUE TITLE
`implicit_return = "deny"` can trigger lint in attribute macro generated code on async functions

### ISSUE THREAD

**Issue by mjbshaw** [rust-lang/rust-clippy#16676]
### Summary

Macros generally shouldn't trigger lints in code they generate (#407). But I've found that if I use `implicit_return = "deny"` and a macro generates that code on an `async fn` this lint triggers. The `async` part causes the `#[tracing::instrument]` macro to generate different code than synchronous, so it's possible `async` isn't relevant to clippy, but I figure I'd mention it explicitly just in case.

### Lint Name

implicit_return

### Reproducer

I tried this code:

**Cargo.toml:**

```
[package]
name = "clippy-implicit-return-instrument-repro"
edition = "2024"

[dependencies]
tracing = { version = "0.1.44", features = ["attributes"] }

[lints.clippy]
implicit_return = "deny"
```

**src/lib.rs:**

```rust
#[tracing::instrument]
pub async fn run() {
}
```

I saw this happen:

```
error: missing `return` statement
 --> src/lib.rs:1:1
  |
1 | #[tracing::instrument]
  | ^^^^^^^^^^^^^^^^^^^^^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.94.0/index.html#implicit_return
  = note: requested on the command line with `-D clippy::implicit-return`
  = note: this error originates in the attribute macro `tracing::instrument` (in Nightly builds, run with -Z macro-backtrace for more info)
```

I expected to see this happen:

Lint should not fire in macro-generated code (see #407).

### Version

```text
rustc 1.94.0 (4a4ef493e 2026-03-02)
binary: rustc
commit-hash: 4a4ef493e3a1488c6e321570238084b38948f6db
commit-date: 2026-03-02
host: aarch64-apple-darwin
release: 1.94.0
LLVM version: 21.1.8
```
